### PR TITLE
Added compile fixture

### PR DESCRIPTION
**Pull request by markkraay**
This MR adds a compile fixture to our integration test suite so that we can test ops both lazily and using the compiler API. Addresses #256.

### CODE REVIEW

File: tripy/tests/integration/conftest.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import pytest
+
+import tripy as tp
+
+@pytest.fixture(params=['compile', 'lazy'])
+def mode(request):
+    return request.param
+
+@pytest.fixture
+def compile_fixture(mode):
```
[the remark this answers is not in the document]
Comment (markkraay) [reply tripy/tests/integration/conftest.py:27]: Yes, I am in the process of implementing `compile_fixture` inside of the tests. I hope to port most of the ops to use this feature in this PR.

**Review by markkraay — COMMENTED** [review]
**`test_shape_tensor[compile]`**
```python
>       raise TripyException(msg) from None
E       tripy.common.exception.TripyException: 
E       
E       --> /tripy/tests/integration/conftest.py:45 in wrapper()
E             |
E          45 |             compiled_func = compiler.compile(*compile_args, **compile_kwargs)
E             | 
E       
E       MTRTException: InternalError: failed to run compilation on module with symbol name: ins_shape_outs_t10_3
E       
E       Additional context:
E       Traceback (most recent call last):
E         File "/tripy/tripy/backend/mlir/compiler.py", line 100, in compile
E           executable = compiler.compiler_stablehlo_to_executable(
E       mlir_tensorrt.runtime._mlir_libs._api.MTRTException: InternalError: failed to run compilation on module with symbol name: ins_shape_outs_t10_3
E       .
E           Error Code: 2: Assertion memSize >= 0 failed. region: result0 dims: [(# 0 (VALUE arg0)),(# 1 (VALUE arg0))]
E           [blockChooser.cpp::getRegionBlockSize::838] Error Code 2: Internal Error (Assertion memSize >= 0 failed. region: result0 dims: [(# 0 (VALUE arg0)),(# 1 (VALUE arg0))])
E           (t10)error: failed to translate function 'tensorrt_cluster' to a TensorRT engine
E       
E           This error occured while trying to compile the following FlatIR expression:
E                 |
E                 | t10: [rank=(2), shape=((-1, -1)), dtype=(float32), loc=(gpu:0)] = DynamicBroadcastOp(t_inter2, shape, broadcast_dim=[])
E                 | 
E       
E       
E           Note: This originated from the following expression:
E       
E           --> /tripy/tripy/frontend/trace/ops/fill.py:127 in full()
E                 |
E             127 |     return full_impl(shape, value, dtype, output_rank)
E                 | 
E       
E           Required from:
E           --> /tripy/tripy/backend/compiler_api.py:524 in compile()
E                 |
E             524 |         trace_outputs = utils.make_list(self.func(*new_args, **new_kwargs))
E                 | 
E       
E           --> /tripy/tests/integration/conftest.py:45 in wrapper()
E                 |
E              45 |             compiled_func = compiler.compile(*compile_args, **compile_kwargs)
E                 |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ --- required from here
E       
E           Input 0:
E       
E           --> /tripy/tripy/frontend/trace/ops/fill.py:127 in full()
E                 |
E             127 |     return full_impl(shape, value, dtype, output_rank)
E                 | 
E       
E           --> /tripy/tripy/backend/compiler_api.py:502 in process_arg()
E                 |
E             502 |                 tensor = full(shape=arg.shape_bounds.opt, value=init_value, dtype=arg.dtype)
E                 |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ --- required from here
E       
E           --> /tripy/tripy/backend/compiler_api.py:514 in compile()
E                 |
E             514 |             new_args.append(process_arg(name, arg))
E                 |                             ^^^^^^^^^^^^^^^^^^^^^^ --- required from here
E       
E           --> /tripy/tests/integration/conftest.py:45 in wrapper()
E                 |
E              45 |             compiled_func = compiler.compile(*compile_args, **compile_kwargs)
E                 |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ --- required from here

tripy/common/exception.py:195: TripyException
-------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------
==== Trace IR ====
t0 = storage(data=[2, 3], shape=(2,), dtype=int32, device=gpu:0)
t2 = fill(t0, value=1, output_rank=2, dtype=int32)
t3 = cast(t2, dtype=float32)
t4 = shape(t3)
outputs:
    t4: [rank=(1), dtype=(int32), loc=(gpu:0)]

==== Flat IR ====
t0: [rank=(1), shape=((2,)), dtype=(int32), loc=(gpu:0)] = ConstantOp(data=[2, 3])
t_inter3: [rank=(0), dtype=(int32), loc=(gpu:0)] = ConstantOp(data=1)
t2: [rank=(2), dtype=(int32), loc=(gpu:0)] = DynamicBroadcastOp(t_inter3, t0, broadcast_dim=[])
t3: [rank=(2), dtype=(float32), loc=(gpu:0)] = ConvertOp(t2)
t_inter6: [rank=(0), shape=(()), dtype=(int32), loc=(gpu:0)] = GetDimensionSizeOp(t3, dim=0)
t_inter7: [rank=(1), shape=([1]), dtype=(int32), loc=(<class 'tripy.common.device.device'>)] = ConstantOp(data=[1])
t_inter5: [rank=(1), shape=((1,)), dtype=(int32), loc=(gpu:0)] = DynamicReshapeOp(t_inter6, t_inter7)
t_inter9: [rank=(0), shape=(()), dtype=(int32), loc=(gpu:0)] = GetDimensionSizeOp(t3, dim=1)
t_inter10: [rank=(1), shape=([1]), dtype=(int32), loc=(<class 'tripy.common.device.device'>)] = ConstantOp(data=[1])
t_inter8: [rank=(1), shape=((1,)), dtype=(int32), loc=(gpu:0)] = DynamicReshapeOp(t_inter9, t_inter10)
t4: [rank=(1), dtype=(int32), loc=(gpu:0)] = ConcatenateOp(t_inter5, t_inter8, dim=0)
outputs:
    t4: [rank=(1), dtype=(int32), loc=(gpu:0)]

==== MLIR ====
module @outs_t4_1 {
  func.func @main() -> tensor<2xi32> {
    %c = stablehlo.constant dense<[2, 3]> : tensor<2xi32>
    %c_0 = stablehlo.constant dense<1> : tensor<i32>
    %0 = stablehlo.dynamic_broadcast_in_dim %c_0, %c, dims = [] : (tensor<i32>, tensor<2xi32>) -> tensor<?x?xi32>
    %1 = stablehlo.convert %0 : (tensor<?x?xi32>) -> tensor<?x?xf32>
    %2 = stablehlo.get_dimension_size %1, dim = 0 : (tensor<?x?xf32>) -> tensor<i32>
    %c_1 = stablehlo.constant dense<1> : tensor<1xi32>
    %3 = stablehlo.reshape %2 : (tensor<i32>) -> tensor<1xi32>
    %4 = stablehlo.get_dimension_size %1, dim = 1 : (tensor<?x?xf32>) -> tensor<i32>
    %c_2 = stablehlo.constant dense<1> : tensor<1xi32>
    %5 = stablehlo.reshape %4 : (tensor<i32>) -> tensor<1xi32>
    %6 = stablehlo.concatenate %3, %5, dim = 0 : (tensor<1xi32>, tensor<1xi32>) -> tensor<2xi32>
    return %6 : tensor<2xi32>
  }
}


==== Trace IR ====
inputs:
    shape: [rank=(1), shape=([2]), dtype=(int32), loc=(gpu:0)]
t10 = fill(shape, value=5.0, output_rank=2, dtype=float32)
outputs:
    t10: [rank=(2), dtype=(float32), loc=(gpu:0)]

==== Flat IR ====
inputs:
    shape: [rank=(1), shape=([2]), dtype=(int32), loc=(gpu:0)]
t_inter2: [rank=(0), dtype=(float32), loc=(gpu:0)] = ConstantOp(data=5.0)
t10: [rank=(2), dtype=(float32), loc=(gpu:0)] = DynamicBroadcastOp(t_inter2, shape, broadcast_dim=[])
outputs:
    t10: [rank=(2), dtype=(float32), loc=(gpu:0)]

==== MLIR ====
module @ins_shape_outs_t10_3 {
  func.func @main(%arg0: tensor<2xi32>) -> tensor<?x?xf32> {
    %cst = stablehlo.constant dense<5.000000e+00> : tensor<f32>
    %0 = stablehlo.dynamic_broadcast_in_dim %cst, %arg0, dims = [] : (tensor<f32>, tensor<2xi32>) -> tensor<?x?xf32>
    return %0 : tensor<?x?xf32>
  }
}

```

**Comment by markkraay** [comment]
@pranavm-nvidia I will continue including details abt. failing test cases + MLIR dumps as I continue to enable `compile_fixture` for more tests. Let me know if there's any additional information / formatting that would be helpful :)

**Comment by pranavm-nvidia** [comment]
> @pranavm-nvidia I will continue including details abt. failing test cases + MLIR dumps as I continue to enable `compile_fixture` for more tests. Let me know if there's any additional information / formatting that would be helpful :)

Could you also include the name of the failing test?

File: tripy/tests/integration/conftest.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# limitations under the License.
+#
+
+import pytest
+
+import tripy as tp
+
+
+@pytest.fixture(params=["compile", "eager"])
+def mode(request):
```
Comment (pranavm-nvidia) [inline tripy/tests/integration/conftest.py:24]: Do we need the separate fixture? Couldn't we parametrize `compile_fixture` directly?

File: tripy/tests/integration/conftest.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# limitations under the License.
+#
+
+import pytest
+
+import tripy as tp
+
+
+@pytest.fixture(params=["compile", "eager"])
+def compile_fixture(request):
```
Comment (yizhuoz004) [inline tripy/tests/integration/conftest.py:24]: https://github.com/NVIDIA/TensorRT-Incubator/issues/102#issuecomment-2293967447

`compile_fixture` should be able to specify which args are constants / parameters. For Q/DQ, scale should be Parameters.

**Comment by yizhuoz004** [comment]
I do not see a lot of value in testing so many low-level APIs in compile mode. What is the time impact on L0?

**Comment by pranavm-nvidia** [comment]
Closing in favor of https://github.com/NVIDIA/TensorRT-Incubator/pull/387

**Pull request closed without merging**